doc/plugin_tutorial/tuto1/src/g_tuto1.mlg

```diff
@@ -16,141 +16,278 @@ open Stdarg
+ * It is important to NEVER use the empty environment or Evd.empty;
+ * if you do, you will get confusing errors.
+ *
+ * NOTE ON STATE: It is important to use the evar_map that is returned to you.
+ * Otherwise, you may get cryptic errors later in your plugin.
+ * For example, you may get universe inconsistency errors.
+ * In general, if a function returns an evar_map to you, that's the one
+ * you want to thread through the rest of your command.
+ *)
+VERNAC COMMAND EXTEND Intern CLASSIFIED AS QUERY
+| [ "Intern" constr(e) ] ->
+   {
+     let env = Global.env () in (* use this; never use empty *)
+     let sigma = Evd.from_env env in (* use this; never use empty *)
+     let debug sigma = Termops.pr_evar_map ~with_univs:true None env sigma in
+     Feedback.msg_notice (strbrk "State before intern: " ++ debug sigma);
+     let (sigma, t) = Constrintern.interp_constr_evars env sigma e in
+     Feedback.msg_notice (strbrk "State after intern: " ++ debug sigma);
+     let print t = Printer.pr_econstr_env env sigma t in
+     Feedback.msg_notice (strbrk "Interned: " ++ print t)
+   }
 END
 
-(* The next step is to make something of parsed expression.
-   Interesting information in interp/constrintern.mli *)
-
-(* There are several phases of transforming a parsed expression into
-  the final internal data-type (constr).  There exists a collection of
-  functions that combine all the phases *)
-
-VERNAC COMMAND EXTEND TakingConstr2 CLASSIFIED AS QUERY
-| [ "Cmd2" constr(e) ] ->
-  { let _ = Constrintern.interp_constr
-    (Global.env())
-    (* Make sure you don't use Evd.empty here, as this does not
-      check consistency with existing universe constraints. *)
-    (Evd.from_env (Global.env())) e in
-    Feedback.msg_notice (strbrk "Cmd2 parsed something legitimate") }
+(*** Defining terms ***)
+
+(*
+ * To define a term, we start similarly to our intern functionality,
+ * then we call another function. We define this function in
+ * the Simple_declare module.
+ *
+ * The line #[ poly = polymorphic ] says that this command accepts
+ * polymorphic attributes.
+ *
+ * The packed_declare_definition function returns the reference
+ * that was defined. This reference will be present in the new environment.
+ * If you want to refer to it later in your plugin, you must use an
+ * updated environment and the constructed reference.
+ *
+ * Note since we are now defining a term, we must classify this
+ * as a side-effect (CLASSIFIED AS SIDEFF).
+ *)
+VERNAC COMMAND EXTEND MyDefine CLASSIFIED AS SIDEFF
+| #[ poly = polymorphic ] [ "MyDefine" ident(i) ":=" constr(e) ] ->
+   {
+     let env = Global.env () in
+     let sigma = Evd.from_env env in
+     let (sigma, t) = Constrintern.interp_constr_evars env sigma e in
+     let r = Simple_declare.packed_declare_definition ~poly i (sigma, t) in
+     let print r = strbrk "Defined " ++ Printer.pr_global r ++ strbrk "." in
+     Feedback.msg_notice (print r)
```

Comment:
I think usually one would put all this code in a function in a ml file, not just the packed_declare_definition part.
In general mlg files should have minimal ml code in them (so as to have proper tool support such as merlin etc) although in some cases it's too hard to extrude the code so we accept it.

Reply:
That's fair, I just feel like for someone reading a tutorial it's really hard to go to the .ml files and look at all of the definitions too. Is it possible to just add a comment that explains this and call it a day? That is, recommend putting almost all OCaml code in .ml files, but explain that we leave it here so it's easier to browse the tutorial?

Reply:
That makes sense.